readme.md
# MeshCentral-ScriptTask

A script running plugin for the [MeshCentral2](https://github.com/Ylianst/MeshCentral) Project. The plugin supports PowerShell, BAT, and Bash scripts. Windows, MacOS, and Linux endpoints are all supported. PowerShell can be run on any OS that has PowerShell installed, not just Windows.

## Important Note
**This plugin now supports NeDB and MongoDB**

## Installation

 Pre-requisite: First, make sure you have plugins enabled for your MeshCentral installation:
>     "plugins": {
>          "enabled": true
>     },
Restart your MeshCentral server after making this change.

 To install, simply add the plugin configuration URL when prompted:
 `https://raw.githubusercontent.com/ryanblenis/MeshCentral-ScriptTask/master/config.json`

## Features
- Add scripts to a central store
- Run scripts on single or multiple endpoints simultaneously
-- Supports PowerShell, BAT, and Bash scripts
- Review the return status / value of completed scripts
-- Returns text or parses returned JSON for easy viewing
- Schedule scripts for future runs on either a one-time or interval basis (minutes, hourly, daily, weekly)
- View schedules and histories based on either node (endpoint) or script
- Dead job detection with script replay

## Usage Notes
- *Run* schedules and runs the job instantly.
- *Schedule* puts the events into a queue.
- Queues are checked / run every minute.
- Scheduled jobs only show the *next* scheduled job at any given time.
- History is limited to 200 events per node/script in the viewport.
- Historical events that have completed will delete after 90 days.
- Jobs only run when the endpoint agent is online. They do *not* queue to the agent when offline, then run at the specified time.
- Scripts are cached on the clients and verified via hash at runtime for the latest version.

## Variables
- Variables can be assigned to the following scopes: Global, Script, Mesh, and Node
- Variables with the same name will override each other in the above order (ex. Global variable testVariable will be overridden by Script variable testVariable will be overridden by Mesh variable Test will be overridden by Node variable testVariable if they exist and apply to the node running the script)
- Variables found in a script which do not correspond with a defined variable will be replaced with VAR_NOT_FOUND
- Variables are assigned during job dispatch (e.g. when the job is sent to the node), so the latest script and variable values will be accounted for, should they have changed since a job or scheduled job were defined.
- Variables in scripts must be defined by hashtags on both sides and not contain spaces. (E.g. #myVar#)

## Getting Started
Drag and drop some of your favorite admin scripts on the file tree. You'll then be able to run them immediately on endpoints (nodes).

## Upcoming Features
- Take action on the results of a script
- Variable replacement

## Other Information
This is fairly new and in beta. Testing was mostly done in the latest Chrome and Firefox. No attempt was made to be compatible with older IE browsers. Endpoint testing was done on Windows 10 (1903) and OS X 10.13.

## Debugging
The plugin emits log message via the built-in Mesh Central "debug" mechanism. The plugin emits logs
using source "plugins". You can get these logs 2 ways:

### Debugging: Server Log Files

Add "plugins" to your config.json file under settings.log:

```JSON
{
  "$schema": "http://info.meshcentral.com/downloads/meshcentral-config-schema.json",
  ...
  "settings": {
    ...
    "log": [
        ...
        "plugins"
    ]
  }
```

The scripttask plugin logs will now appear in the server logs.

### Debugging: Mesh Central Web Portal

The Mesh Central UI offers a "tracing" mechanism under My Server > Trace. With the out-of-box code,
Mesh Central does not offer the ability to capture "plugins" log records using the built-in
interface. That is to say, the sources you can capture are hard-coded into the user interface :(

Fortunately, much of JavaScript which drives the UI is accessible on the DOM Window object. You can
use the following code in Chrome DevTools to enable tracing of the "plugins" source:

```JavaScript
((sources) => {
    serverTraceSources = sources;
	meshserver.send({ action: 'traceinfo', traceSources: sources});
	QH('p41traceStatus', EscapeHtml(sources.join(', ')));
})([ 'plugins' ]);
```

This code:

* Updates the serverTraceSources global variable
* Sends a message to the server instructing the server to send "plugins" source events to you
* Updates the UI to indicate you are tracing these events

NOTE: you may add additional sources if you want to trace other sources in addition to "plugins"

## Screenshots
![Device Page](https://user-images.githubusercontent.com/1929277/71248033-f4519b00-22e7-11ea-9aa6-ef22e0b9fdb2.png)
![Script Editor](https://user-images.githubusercontent.com/1929277/71248034-f4519b00-22e7-11ea-8ab4-ccad3e959a1a.png)
![Schedule Screen](https://user-images.githubusercontent.com/1929277/71248469-e7817700-22e8-11ea-9121-a215de160e0e.png)



changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [Known Issues]
- None. Please feel free to submit an issue via [GitHub](https://github.com/ryanblenis/MeshCentral-ScriptTask) if you find anything.

## [0.0.20] - 2023-02-06
### Added
- Added `-ExecutionPolicy Bypass` flag to powershell invocation

## [0.0.18] - 2022-03-12
### Fixed
- Compatibility with MeshCentral => 0.9.98 (promise requirement removed from MeshCentral, nedb moved to @yetzt/nedb)

## [0.0.17] - 2021-10-07
### Fixed
- Compatibility with MeshCentral > 0.9.7

## [0.0.16] - 2021-05-18
### Fixed
- Drag/drop scripts between folders

## [0.0.15] - 2020-03-07
### Fixed
- Prevent race condition with conflicting plugins changing the app views directory by using absolute paths

## [0.0.14] - 2020-02-28
### Added
- Add new script from ScriptTask plugin without dragging/uploading an existing script
- Variables introduced. Supports global, script, mesh, and node level. Overridden in that order.

## [0.0.13] - 2020-02-15
### Fixed
- Folder rename function now working as intended
- Removed ability to "edit" folders as if they were a script

## [0.0.12] - 2020-02-14
### Fixed
- Rename function broke when filetype tags were added

## [0.0.11] - 2020-01-24
### Added
- Advanced selection of nodes based on mesh and tag membership

## [0.0.10] - 2020-01-02
### Fixed
- Removed logging to MeshCentral events DB for each script run / refresh event (causing events file to grow quickly and crash NeDB instances)

## [0.0.9] - 2019-12-24
### Fixed
- Native Powershell scripts now run correctly on more versions of Linux/MacOS

## [0.0.8] - 2019-12-23
### Fixed
- Force strip carriage returns from bash scripts

## [0.0.7] - 2019-12-23
### Fixed
- Some versions of Linux don't default the agents root to the MeshAgent's directory. Added path detection for a safe temp zone. Thanks /u/Reverent for the assist in debugging that one!

## [0.0.6] - 2019-12-23
### Added
- Status return of a script as success/failure if the script does not produce output
- Endpoint debugging toggle via console (plugin scripttask debug)
### Fixed
- Made some endpoint calls a little safer with try/catch to stop errors from killing the process

## [0.0.5] - 2019-12-22
### Added
- NeDB support
### Fixed
- CSS call looking for mapping file removed
- Date/time select on schedule screen might not have had the correct mimetype, thus not appeared in some instances

## [0.0.4] - 2019-12-22
### Added
- Script type identifier to the script tree view
### Fixed
- Prevent catch-up jobs from running every minute on long running schedules when a node is offline for a period of time.
- Check for scheduled jobs 1 minute in the future to prevent schedules running 1 minute behind intended time.

## [0.0.3] - 2019-12-20
### Fixed
- Typo causing installs to think Mongo is not in use

## [0.0.2] - 2019-12-20
### Added
- Suppressed errors and made a nice error message for those without MongoDB (currently unsupported)

## [0.0.1] - 2019-12-20
### Added
- Released initial version


db.js
/** 
* @description MeshCentral-ScriptTask database module
* @author Ryan Blenis
* @copyright Ryan Blenis 2019
* @license Apache-2.0
*/

"use strict";

/**
 * A job execution instruction
 * @typedef {Object} Job
 * @property {ObjectID} _id
 * @property {'job'} type - type of record (for MongoDB discrimination)
 * @property {number} queueTime - UTC millis when job was queued
 * @property {number} dontQueueUntil - UTC millis of when job is eligble to be queued
 * @property {number} dispatchTime - UTC millis when the job was dispatched to client
 * @property {number} completeTime - UTC millis when the job was completed
 * @property {string} node - mesh node ID the job is to run on
 * @property {string} scriptId - id of the script to execute
 * @property {string} scriptName - name of the script to run
 * @property {unknown|null} replaceVars - variable replacement collection
 * @property {string|null} returnVal - result of the job (i.e. script stdout once complete)
 * @property {string|null} errorVal - error output of the job (i.e. script stderr once complete)
 * @property {unknown} returnAct - unknown purpose
 * @property {string} runBy - username that invoked the job
 * @property {ObjectID} jobSchedule - reference to job schedule
 * @property {jobState} state - state of the job
 */

/**
 * A job execution schedule
 * @typedef {Object} JobSchedule
 * @property {ObjectID} _id - mongo ID
 * @property {'jobSchedule'} type - type discriminator
 * @property {string} scriptId - id of the script
 * @property {string} node - node the script is scheduled on
 * @property {string} scheduledBy - who made the schedule
 * @property {'once'|'minutes'|'hourly'|'daily'|'weekly'} recur - recurrence basis
 * @property {number} interval - recurrence interval
 * @property {unknown} daysOfWeek - which days of the week to run on
 * @property {number} startAt - UTC millis of when the schedule begins
 * @property {number} endAt - UTC millis of when the schedule ends
 * @property {number} lastRun - UTC millis of the last execution of this schedule
 * @property {number} nextRun - UTC millis of the next time to run the schedule
 * @property {string} target - scheduling tartget (e.g. "Jane Doe's Laptop" or "Mesh: IT-Test")
 */

/**
 * A mesh-based job execution schedule
 * @typedef {Object} MeshJobSchedule
 * @property {ObjectID} _id - mongo ID
 * @property {'meshJobSchedule'} type - type discriminator
 * @property {string} scriptId - id of the script
 * @property {string} mesh - mesh the script is scheduled on
 * @property {string} scheduledBy - who made the schedule
 * @property {'once'|'minutes'|'hourly'|'daily'|'weekly'} recur - recurrence basis
 * @property {number} interval - recurrence interval
 * @property {unknown} daysOfWeek - which days of the week to run on
 * @property {number} startAt - UTC millis of when the schedule begins
 * @property {number} endAt - UTC millis of when the schedule ends
 * @property {number} lastRun - UTC millis of the last execution of this schedule
 * @property {number} nextRun - UTC millis of the next time to run the schedule
 * @property {string} target - scheduling tartget (e.g. "Jane Doe's Laptop" or "Mesh: IT-Test")
 */

/**
 * A script
 * @typedef {Object} Script
 * @property {ObjectID} _id - mongo ID
 * @property {'script'} type - type discriminator
 * @property {string} path - virtual folder path
 * @property {string} name - script name
 * @property {string} content - script contents
 * @property {string} contentHash - script hash
 * @property {'ps1'|'bat'|'bash'} filetype - type of script
 */

const jobState = require('./jobState');

var Datastore = null;
var formatId = null;

module.exports.CreateDB = function(meshserver) {
    var obj = {};
    var NEMongo = require(__dirname + '/nemongo.js');
    obj.dbVersion = 1;
    
    obj.initFunctions = function () {
        obj.updateDBVersion = function(new_version) {
          return obj.scriptFile.updateOne({type: "db_version"}, { $set: {version: new_version} }, {upsert: true});
        };
        
        obj.getDBVersion = function() {
            return new Promise(function(resolve, reject) {
                obj.scriptFile.find( { type: "db_version" } ).project( { _id: 0, version: 1 } ).toArray(function(err, vers){
                    if (vers.length == 0) resolve(1);
                    else resolve(vers[0]['version']);
                });
            });
        };

        obj.addScript = function(name, content, path, filetype) {
            if (path == null) path = "Shared"
            if (filetype == 'bash') content = content.split('\r\n').join('\n').split('\r').join('\n');
            var sObj = { 
                type: 'script',
                path: path,
                name: name,
                content: content,
                contentHash: require('crypto').createHash('sha384').update(content).digest('hex'),
                filetype: filetype
            };
            return obj.scriptFile.insertOne(sObj);
        };
        
        obj.addFolder = function(name, path) {
            var sObj = {
                type: 'folder',
                path: path,
                name: name
            };
            return obj.scriptFile.insertOne(sObj);
        };
        
        obj.getScriptTree = function() {
            return obj.scriptFile.find(
                { type:
                    { $in: [ 'script', 'folder' ] }
                }
            ).sort(
                { path: 1, type: 1, name: 1 }
            ).project(
                { name: 1, path: 1, type: 1, filetype: 1 }
            ).toArray();
        };
        
        obj.update = function(id, args) {
            id = formatId(id);
            if (args.type == 'script' && args.content !== null) { 
                if (args.filetype == 'bash') {
                    args.content = args.content = split('\r\n').join('\n').split('\r').join('\n');
                }
                args.contentHash = require('crypto').createHash('sha384').update(args.content).digest('hex');
            }
            return obj.scriptFile.updateOne( { _id: id }, { $set: args } );
        };

        obj.delete = function(id) {
            id = formatId(id);
            return obj.scriptFile.deleteOne( { _id: id } );
        };

        obj.deleteByPath = function(path) {
          return obj.scriptFile.deleteMany( { path: path, type: { $in: ['script', 'folder'] } } );
        };

        obj.deleteSchedulesForScript = function(id) {
            id = formatId(id);
            return obj.scriptFile.deleteMany( { type: 'jobSchedule', scriptId: id } );
        };

        obj.getByPath = function(path) {
          return obj.scriptFile.find( { type: { $in: [ 'script', 'folder' ] }, path: path }).toArray();
        };

        obj.get = function(id) {
            if (id == null || id == 'null') return new Promise(function(resolve, reject) { resolve([]); });
            id = formatId(id);
            return obj.scriptFile.find( { _id: id } ).toArray();
        };

        obj.addJob = function(passedObj) {
          var nowTime = Math.floor(new Date() / 1000);
          var defaultObj = { 
              type: 'job',
              queueTime: nowTime,
              dontQueueUntil: nowTime,
              dispatchTime: null,
              completeTime: null,
              node: null,
              scriptId: null,
              scriptName: null, // in case the original reference is deleted in the future
              replaceVars: null,
              returnVal: null,
              errorVal: null,
              returnAct: null,
              runBy: null,
              jobSchedule: null,
              state: jobState.CREATED,
          };
          var jObj = {...defaultObj, ...passedObj};
          
          if (jObj.node == null || jObj.scriptId == null) { console.log('PLUGIN: SciptTask: Could not add job'); return false; }
          
          return obj.scriptFile.insertOne(jObj);
        };

        obj.addMeshJobSchedule = function(schedObj) {
            schedObj.type = 'meshJobSchedule';

            if(schedObj.mesh == null) {
                throw new Error(`failed to add mesh job schedule: property 'mesh' is nullish`);
            }

            if(schedObj.scriptId == null) {
                throw new Error(`failed to add mesh job schedule: property 'scriptId' is nullish`);
            }

            return obj.scriptFile.insertOne(schedObj);
        };

        obj.addJobSchedule = function(schedObj) {
            schedObj.type = 'jobSchedule';

            if (schedObj.node == null || schedObj.scriptId == null) {
                console.log('PLUGIN: SciptTask: Could not add job schedule');
                
                return false;
            }

            return obj.scriptFile.insertOne(schedObj);
        };

        obj.removeJobSchedule = function (id) {
            return obj.delete(id);
        };

        /**
         * Returns JobSchedule objects which are overdue (nextRun <= now) and for which the current
         * time is within the execution window (endAt <= now).
         * 
         * If scheduleId is provided, limits results to that specific schedule.
         * 
         * @param {string|undefined} scheduleId - mongo DB _id for schedule
         * @param {'jobSchedule'|'meshJobSchedule'|undefined} - type of schedule to search. Default: jobSchedule
         * @returns {Promise.<Array.<JobSchedule>>}
         */
        obj.getSchedulesDueForJob = function(scheduleId, type) {
            var nowTime = Math.floor(new Date() / 1000);
            var scheduleIdLimiter = {};
            if (scheduleId != null) {
                scheduleIdLimiter._id = scheduleId;
            }

            return obj.scriptFile.find( { 
                type: type || 'jobSchedule',
                $or: [
                    { endAt: null },
                    { endAt: { $lte: nowTime } }
                ],
                $or: [
                    { nextRun: null }, 
                    { nextRun: { $lte: nowTime } }
                ],
                ...scheduleIdLimiter
            }).toArray();
        };

        obj.deletePendingJobsForNode = function(node) {
            return obj.scriptFile.deleteMany({ 
                type: 'job', 
                node: node,
                completeTime: null,
            });
        };

        obj.getPendingJobs = function(nodeScope) {
          if (nodeScope == null || !Array.isArray(nodeScope)) {
            return false;
          }
          // return jobs that has online nodes and queue time requirements have been met
          return obj.scriptFile.find( { 
              type: 'job', 
              node: { $in: nodeScope },
              completeTime: null,
              //dispatchTime: null,
              $or: [
                  { dontQueueUntil: null }, 
                  { dontQueueUntil: { $lte: Math.floor(new Date() / 1000) } }
              ]
          }).toArray();
        };

        obj.getJobNodeHistory = function(nodeId) {
            return obj.scriptFile.find( { 
                type: 'job', 
                node: nodeId,
            }).sort({ queueTime: -1 }).limit(200).toArray();
        };

        /**
         * Returns an array of Job for scriptId sorted desc by completeTime, then by queueTime.
         * Limits results to 200 items.
         * 
         * @param {string} scriptId 
         * @returns {Promise.<Array.<Job>>}
         */
        obj.getJobScriptHistory = function(scriptId) {
            return obj.scriptFile
                .find({ 
                    type: 'job', 
                    scriptId: scriptId,
                })
                .sort({ completeTime: -1, queueTime: -1 })
                .limit(200)
                .toArray();
        };

        obj.updateScriptJobName = function(scriptId, scriptName) {
            return obj.scriptFile.updateMany({ type: 'job', scriptId: scriptId }, { $set: { scriptName: scriptName } });    
        };

        /**
         * 
         * @param {string} scriptId 
         * @returns {Promise.<Array.<JobSchedule|MeshJobSchedule>>}
         */
        obj.getJobSchedulesForScript = function(scriptId) {
            return obj.scriptFile.find({
                type: { '$in': ['jobSchedule', 'meshJobSchedule'] },
                scriptId: scriptId
            }).toArray();
        };

        obj.getJobSchedulesForNode = function (nodeId) {
            return obj.scriptFile.find( { type: 'jobSchedule', node: nodeId } ).toArray();
        };

        /**
         * Returns incomplete jobs associated with a JobSchedule or MeshJobSchedule
         * @param {string} schedId 
         * @param {string|undefined} nodeId 
         * @returns {Array.<Job>}
         */
        obj.getIncompleteJobsForSchedule = function (schedId, nodeId) {
            return obj.scriptFile.find({
                type: 'job',
                state: 1, //scheduled
                jobSchedule: schedId,
                completeTime: null,
                ...(nodeId ? { node: nodeId } : undefined),
            }).toArray();
        };

        obj.deletePendingJobsForSchedule = function (schedId) {
            return obj.scriptFile.deleteMany( { type: 'job', jobSchedule: schedId, completeTime: null } );
        };

        obj.deleteOldHistory = function() {
            var nowTime = Math.floor(new Date() / 1000);
            var oldTime = nowTime - (86400 * 90); // 90 days
            return obj.scriptFile.deleteMany( { type: 'job', completeTime: { $lte: oldTime } } );
        };

        obj.addVariable = function(name, scope, scopeTarget, value) {
            var vObj = { 
                type: 'variable',
                name: name,
                scope: scope,
                scopeTarget: scopeTarget,
                value: value
            };
            return obj.scriptFile.insertOne(vObj);
        };

        obj.getVariables = function(limiters) {
            if (limiters != null) {
                var find = { 
                    type: 'variable',
                    name: { $in: limiters.names },
                    $or: [ 
                        { scope: 'global' },
                        { $and: [
                            { scope: 'script' },
                            { scopeTarget: limiters.scriptId }
                          ]
                        },
                        { $and: [
                            { scope: 'mesh' },
                            { scopeTarget: limiters.meshId }
                          ]
                        },
                        { $and: [
                            { scope: 'node' },
                            { scopeTarget: limiters.nodeId }
                          ]
                        }
                    ]
                };
                return obj.scriptFile.find(find).sort({ name: 1 }).toArray();
            }
            else {
                return obj.scriptFile.find({ type: 'variable' }).sort({ name: 1 }).toArray();
            }
        };

        obj.checkDefaults = function() {
            obj.scriptFile.find( { type: 'folder', name: 'Shared', path: 'Shared' } ).toArray()
            .then(found => {
              if (found.length == 0) obj.addFolder('Shared', 'Shared');
            })
            .catch(e => { console.log('PLUGIN: ScriptTask: Default folder check failed. Error was: ', e); });
        };
        
        obj.checkDefaults();
    };
    
    if (meshserver.args.mongodb) { // use MongDB
      require('mongodb').MongoClient.connect(meshserver.args.mongodb, { useNewUrlParser: true, useUnifiedTopology: true }, function (err, client) {
          if (err != null) { console.log("Unable to connect to database: " + err); process.exit(); return; }
          
          var dbname = 'meshcentral';
          if (meshserver.args.mongodbname) { dbname = meshserver.args.mongodbname; }
          const db = client.db(dbname);
          
          obj.scriptFile = db.collection('plugin_scripttask');
          obj.scriptFile.indexes(function (err, indexes) {
              // Check if we need to reset indexes
              var indexesByName = {}, indexCount = 0;
              for (var i in indexes) { indexesByName[indexes[i].name] = indexes[i]; indexCount++; }
              if ((indexCount != 6) || (indexesByName['ScriptName1'] == null) || (indexesByName['ScriptPath1'] == null) || (indexesByName['JobTime1'] == null) || (indexesByName['JobNode1'] == null) || (indexesByName['JobScriptID1'] == null)) {
                  // Reset all indexes
                  console.log('Resetting plugin (ScriptTask) indexes...');
                  obj.scriptFile.dropIndexes(function (err) {
                      obj.scriptFile.createIndex({ name: 1 }, { name: 'ScriptName1' });
                      obj.scriptFile.createIndex({ path: 1 }, { name: 'ScriptPath1' });
                      obj.scriptFile.createIndex({ queueTime: 1 }, { name: 'JobTime1' });
                      obj.scriptFile.createIndex({ node: 1 }, { name: 'JobNode1' });
                      obj.scriptFile.createIndex({ scriptId: 1 }, { name: 'JobScriptID1' });
                  }); 
              }
          });
          
          
          if (typeof require('mongodb').ObjectID == 'function') {
              formatId = require('mongodb').ObjectID;
          } else {
              formatId = require('mongodb').ObjectId;
          }
          obj.initFunctions();
    });  
    } else { // use NeDb
        Datastore = require('@yetzt/nedb');
        if (obj.scriptFilex == null) {
            obj.scriptFilex = new Datastore({ filename: meshserver.getConfigFilePath('plugin-scripttask.db'), autoload: true });
            obj.scriptFilex.persistence.setAutocompactionInterval(40000);
            obj.scriptFilex.ensureIndex({ fieldName: 'name' });
            obj.scriptFilex.ensureIndex({ fieldName: 'path' });
            obj.scriptFilex.ensureIndex({ fieldName: 'queueTime' });
            obj.scriptFilex.ensureIndex({ fieldName: 'node' });
            obj.scriptFilex.ensureIndex({ fieldName: 'scriptId' });
        }
        obj.scriptFile = new NEMongo(obj.scriptFilex);
        formatId = function(id) { return id; };
        obj.initFunctions();
    }
    
    return obj;
}